### PR TITLE
element wise operators support dynamic shape input and broadcast

### PR DESCRIPTION
**Pull request by uni19**
# Description

## Type of change
- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by narendasan** [comment]
@uni19 Do you have testcases for this patch?

**Comment by uni19** [comment]
> @uni19 Do you have testcases for this patch?

updated

**Comment by narendasan** [comment]
@peri044 Can you take a look at this?

**Comment by narendasan** [comment]
We merged a bunch of linting stuff, can you rebase @uni19? Thanks for sticking with this!

**Comment by uni19** [comment]
> We merged a bunch of linting stuff, can you rebase @uni19? Thanks for sticking with this!

done

File: core/conversion/converters/impl/element_wise.cpp [hunk 1]
```diff
@@ -9,246 +9,326 @@ namespace converters {
...
+            auto s1 =
+                ctx->net
+                    ->addElementWise(*selfShape, *selfShapeMX,
+                                     nvinfer1::ElementWiseOperation::kPROD)
+                    ->getOutput(0);
+            auto s2 = ctx->net
+                          ->addElementWise(*s1, *otherShapeTX,
+                                           nvinfer1::ElementWiseOperation::kSUM)
+                          ->getOutput(0);
+
```
Comment (peri044) [inline core/conversion/converters/impl/element_wise.cpp:47]: General comment: The formatting seems a bit odd. Can you try [running linter using these instructions](https://github.com/NVIDIA/TRTorch/blob/master/CONTRIBUTING.md#coding-guidelines) and format it to match the style guidelines?

File: core/conversion/converters/impl/element_wise.cpp [hunk 2]
```diff
@@ -9,246 +9,326 @@ namespace converters {
...
+            auto selfShape = ctx->net->addShape(*self)->getOutput(0);
+            auto s1 =
+                ctx->net
+                    ->addElementWise(*selfShape, *selfShapeMX,
+                                     nvinfer1::ElementWiseOperation::kPROD)
+                    ->getOutput(0);
+            auto s2 = ctx->net
+                          ->addElementWise(*s1, *otherShapeTX,
+                                           nvinfer1::ElementWiseOperation::kSUM)
+                          ->getOutput(0);
```
Comment (peri044) [inline core/conversion/converters/impl/element_wise.cpp:46]: Can you put some brief comments on what exactly is happening in this snippet for easy reference ?

File: core/conversion/converters/impl/element_wise.cpp [hunk 3]
```diff
@@ -9,246 +9,326 @@ namespace converters {
...
+                    otherShapeT[start + idx] = otherDim[idx];
+                } else {
+                    otherShapeT[start + idx] = 0;
+                    selfShapeM[start + idx] = 1;
+                }
+            }
+            auto otherShapeTX = tensor_to_const(ctx, otherShapeT);
+            auto selfShapeMX = tensor_to_const(ctx, selfShapeM);
+            auto selfShape = ctx->net->addShape(*self)->getOutput(0);
+            auto s1 =
```
Comment (peri044) [inline core/conversion/converters/impl/element_wise.cpp:38]: Variable names `s1` and `s2` can be renamed maybe to better reflect the function ?

File: tests/core/converters/test_element_wise.cpp [hunk 4]
```diff
@@ -4,19 +4,25 @@
...
+    if (!singleInput) {
+        torch_inputs.push_back(at::randint(1, 5, shape2, {at::kCUDA}));
+    }
+    auto params = trtorch::core::conversion::get_named_params(g->inputs(), {});
+    auto jit_results = trtorch::tests::util::RunGraph(g, params, torch_inputs);
+    
+    std::vector<at::Tensor> trt_inputs;
+    for (auto in : torch_inputs) {
+        trt_inputs.push_back(at::clone(in));
+    }
```
[the remark this answers is not in the document]
Comment (uni19) [reply tests/core/converters/test_element_wise.cpp]: updated

File: tests/core/converters/test_element_wise.cpp [hunk 5]
```diff
@@ -35,58 +41,47 @@ TEST(Converters, ATenAddConvertsCorrectly) {
...
 }
 
 TEST(Converters, ATenPowTensorConvertsCorrectly) {
   const auto graph = R"IR(
        graph(%x.1 : Tensor, %x2.1 : Tensor):
           %3 : Tensor = aten::pow(%x.1, %x2.1)
           return (%3))IR";
-  pointwise_test_helper(graph, false);
+    pointwise_test_helper(graph, false);
+    pointwise_test_helper(graph, false, {3, 4}, {4});
```
[the remark this answers is not in the document]
Comment (uni19) [reply tests/core/converters/test_element_wise.cpp]: fixed

File: core/conversion/converters/impl/element_wise.cpp (hunk 1 above)
Comment (uni19) [reply core/conversion/converters/impl/element_wise.cpp]: updated

File: core/conversion/converters/impl/element_wise.cpp (hunk 2 above)
Comment (uni19) [reply core/conversion/converters/impl/element_wise.cpp]: updated

File: core/conversion/converters/impl/element_wise.cpp (hunk 3 above)
Comment (uni19) [reply core/conversion/converters/impl/element_wise.cpp]: updated

File: core/conversion/converters/impl/element_wise.cpp [hunk 6]
```diff
@@ -14,237 +14,267 @@ nvinfer1::ILayer* add_elementwise(
...
+        }
+      }
+      auto otherStaticShapeMask = tensor_to_const(ctx, thOtherStaticShapeMask);
+      auto otherDynamicShapeMask = tensor_to_const(ctx, thOtherDynamicShapeMask);
+      auto selfShape = ctx->net->addShape(*self)->getOutput(0);
+      // size of dynamic dimension of other need to the same as that of corresponding dimension of self
+      auto otherDynamicShape =
+          ctx->net->addElementWise(*selfShape, *otherDynamicShapeMask, nvinfer1::ElementWiseOperation::kPROD)
+              ->getOutput(0);
+      auto targetOhterShape =
```
Comment (peri044) [inline core/conversion/converters/impl/element_wise.cpp:46]: targetOhterShape -> targetOtherShape  (minor correction in variable name)

**Review by peri044 — COMMENTED** [review]
Thanks for resolving previous comments. I've left a minor correction.
As I ran the tests, every testcase seems to be going into static block. Can you add a testcase which makes use of the dynamic shape code you've added ? Thanks

File: core/conversion/converters/impl/element_wise.cpp (hunk 6 above)
Comment (uni19) [reply core/conversion/converters/impl/element_wise.cpp]: updated

**Comment by uni19** [comment]
> Thanks for resolving previous comments. I've left a minor correction.
> As I ran the tests, every testcase seems to be going into static block. Can you add a testcase which makes use of the dynamic shape code you've added ? Thanks

I don't know how to make it dynamic in unit test, can you kindly help me @peri044  ?

File: core/conversion/converters/impl/element_wise.cpp [hunk 7]
```diff
@@ -43,13 +43,13 @@ nvinfer1::ILayer* add_elementwise(
...
               ->getOutput(0);
-      auto targetOhterShape =
+      auto targetOtherShape =
           ctx->net->addElementWise(*otherDynamicShape, *otherStaticShapeMask, nvinfer1::ElementWiseOperation::kSUM)
               ->getOutput(0);
 
       auto otherShuffle = ctx->net->addShuffle(*other);
       otherShuffle->setName(std::string("Reshape other tensor to have the same nDim as self for " + name).c_str());
-      otherShuffle->setInput(1, *targetOhterShape);
+      otherShuffle->setInput(1, *targetOtherShape);
```
Comment (peri044) [inline core/conversion/converters/impl/element_wise.cpp:52]: How did you test this block of code? The testcases don't seem to be executing this either 
```
if (otherDim.end() != std::find(otherDim.begin(), otherDim.end(), -1)) {
      auto thOtherStaticShapeMask = torch::ones(selfDim.size(), torch::kInt32);
      auto thOtherDynamicShapeMask = torch::zeros(selfDim.size(), torch::kInt32);
      for (size_t start = selfDim.size() - otherDim.size(), idx = 0; idx < otherDim.size(); ++idx) {
        if (-1 != otherDim[idx]) {
          thOtherStaticShapeMask[start + idx] = otherDim[idx];
        } else {
          thOtherStaticShapeMask[start + idx] = 0;
          thOtherDynamicShapeMask[start + idx] = 1;
        }
      }
      auto otherStaticShapeMask = tensor_to_const(ctx, thOtherStaticShapeMask);
      auto otherDynamicShapeMask = tensor_to_const(ctx, thOtherDynamicShapeMask);
      auto selfShape = ctx->net->addShape(*self)->getOutput(0);
      // size of dynamic dimension of other need to the same as that of corresponding dimension of self
      auto otherDynamicShape =
          ctx->net->addElementWise(*selfShape, *otherDynamicShapeMask, nvinfer1::ElementWiseOperation::kPROD)
              ->getOutput(0);
      auto targetOtherShape =
          ctx->net->addElementWise(*otherDynamicShape, *otherStaticShapeMask, nvinfer1::ElementWiseOperation::kSUM)
              ->getOutput(0);

      auto otherShuffle = ctx->net->addShuffle(*other);
      otherShuffle->setName(std::string("Reshape other tensor to have the same nDim as self for " + name).c_str());
      otherShuffle->setInput(1, *targetOtherShape);
      other = otherShuffle->getOutput(0);
```

Comment (uni19) [reply core/conversion/converters/impl/element_wise.cpp:52]: It seem that RunGraphEngineDynamic requires input tensors to have at least 2 dimensions to run in dynamic shape state. Fixed

**Comment by uni19** [comment]
> Thanks for resolving previous comments. I've left a minor correction.
> As I ran the tests, every testcase seems to be going into static block. Can you add a testcase which makes use of the dynamic shape code you've added ? Thanks

updated

**Comment by uni19** [comment]
#223

**Pull request closed without merging**